pipeline/metadata/satellite.py

```diff
@@ -225,14 +236,9 @@ def _add_vantage_point_tags(
     Returns:
       PCollection of measurement rows with tag information added to the ip row
   """
-  # PCollection[Row]
-  vp_tags = (
-      tags | 'filter out received ip tags' >>
-      beam.Filter(_is_vantage_point_tag).with_output_types(Row))
-
   # PCollection[Tuple[DateIpKey,Row]]
   ips_with_metadata = (
-      vp_tags | 'tags key by ips and dates' >>
+      tags | 'tags key by ips and dates' >>
       beam.Map(lambda row: (make_date_ip_key(row), row)) |
       'combine duplicate tags' >>
       beam.CombinePerKey(_merge_dicts).with_output_types(Tuple[DateIpKey, Row]))
```

Comment:
Do you need this step? Why not pass `tags` directly to the CoGroupByKey below? That will already group the tags by key. You can remove this step by adjusting grouped_metadata_and_rows.
This will remove an entire shuffle+merge+serialization step.
It's fine if you prefer to do that in a separate PR.

As a general rule of thumb, delay merging as much as possible. Dealing with partitioned data is more efficient than dealing with merged data.

Reply:
This is combining metadata which have the same ip and date, but separate tags. In this case different resolver tags are stored in separate files, so we could have a situation like

tag1 {"date":"2022-01-01", "ip":"1.2.3.4", "name": "dns-resolver.com"}
tag2 {"date":"2022-01-01", "ip":"1.2.3.4", "country_code": "US"}

This merges them into a single tag. `merge_metadata_with_rows` currently doesn't know how to deal with cases where we have more than one dict of metadata on the right side of the join.

Reply:
`merge_metadata_with_rows` will get a list under `IP_METADATA_PCOLLECTION_NAME` where each entry will be one of the metadata objects. Just iterate over the list and add the metadata to your target row. There's no need to pre-merge the metadata objects. This also prevents the need for a type for the merged annotations.

Reply:
This makes me nervous because `merge_metadata_with_rows` is shared code with hyperquack, but okay. Done.